lib/src/codec.dart

```diff
@@ -96,163 +73,222 @@ class Codec extends StandardMessageCodec {
+    v.useBinaryProtocol = readFromJson<bool>(jsonMap, TxClientOptions.useBinaryProtocol);
+    v.queueMessages = readFromJson<bool>(jsonMap, TxClientOptions.queueMessages);
+    v.echoMessages = readFromJson<bool>(jsonMap, TxClientOptions.echoMessages);
+    v.recover = readFromJson<String>(jsonMap, TxClientOptions.recover);
+    v.environment = readFromJson<String>(jsonMap, TxClientOptions.environment);
+    v.idempotentRestPublishing = readFromJson<bool>(jsonMap, TxClientOptions.idempotentRestPublishing);
+    v.httpOpenTimeout = readFromJson<int>(jsonMap, TxClientOptions.httpOpenTimeout);
+    v.httpRequestTimeout = readFromJson<int>(jsonMap, TxClientOptions.httpRequestTimeout);
+    v.httpMaxRetryCount = readFromJson<int>(jsonMap, TxClientOptions.httpMaxRetryCount);
+    v.realtimeRequestTimeout = readFromJson<int>(jsonMap, TxClientOptions.realtimeRequestTimeout);
+    v.fallbackHosts = readFromJson<List<String>>(jsonMap, TxClientOptions.fallbackHosts);
+    v.fallbackHostsUseDefault = readFromJson<bool>(jsonMap, TxClientOptions.fallbackHostsUseDefault);
+    v.fallbackRetryTimeout = readFromJson<int>(jsonMap, TxClientOptions.fallbackRetryTimeout);
+    v.defaultTokenParams = decodeTokenParams(toJsonMap(jsonMap[TxClientOptions.defaultTokenParams]));
+    v.channelRetryTimeout = readFromJson<int>(jsonMap, TxClientOptions.channelRetryTimeout);
+    v.transportParams = readFromJson<Map<String, String>>(jsonMap, TxClientOptions.transportParams);
+    return v;
+  }
+
+  TokenDetails decodeTokenDetails(Map<String, dynamic> jsonMap){
+    if(jsonMap==null) return null;
+    TokenDetails v = TokenDetails(jsonMap[TxTokenDetails.token]);
+    v.expires = readFromJson<int>(jsonMap, TxTokenDetails.expires);
+    v.issued = readFromJson<int>(jsonMap, TxTokenDetails.issued);
+    v.capability = readFromJson<String>(jsonMap, TxTokenDetails.capability);
+    v.clientId = readFromJson<String>(jsonMap, TxTokenDetails.clientId);
     return v;
   }
 
-  TokenDetails decodeTokenDetails(ReadBuffer buffer){
-    TokenDetails t = TokenDetails(readValue(buffer));
-    t.expires = readValue(buffer) as int;
-    t.issued = readValue(buffer) as int;
-    t.capability = readValue(buffer) as String;
-    t.clientId = readValue(buffer) as String;
-    return t;
+  TokenParams decodeTokenParams(Map<String, dynamic> jsonMap){
+    if(jsonMap==null) return null;
+    TokenParams v = TokenParams();
+    v.capability = readFromJson<String>(jsonMap, TxTokenParams.capability);
+    v.clientId = readFromJson<String>(jsonMap, TxTokenParams.clientId);
+    v.nonce = readFromJson<String>(jsonMap, TxTokenParams.nonce);
+    v.timestamp = readFromJson<DateTime>(jsonMap, TxTokenParams.timestamp);
+    v.ttl = readFromJson<int>(jsonMap, TxTokenParams.ttl);
+    return v;
   }
 
-  AblyMessage decodeAblyMessage(ReadBuffer buffer){
+  AblyMessage decodeAblyMessage(Map<String, dynamic> jsonMap){
+    if(jsonMap==null) return null;
```

Comment:
My gut feeling without looking at how this is called more deeply is that this feels like a null argument exception... or do you document this behaviour somewhere?

Reply:
@tiholic ?

Reply:
@QuintinWillison updated documentation for codec.dart - 237315e9a5b5a65bdc7abe091f4611bd4ef32c77